pkg/morph/event/notary_preparator.go

```diff
@@ -0,0 +1,316 @@
+	// contract
+	if len(nr.MainTransaction.Scripts) != 3 {
+		return nil, errUnexpectedWitnessAmount
+	}
+
+	// alphabet node should handle only notary requests
+	// that have been sent unsigned(by storage nodes) =>
+	// such main TXs should have dummy scripts as an
+	// invocation script
+	//
+	// this check prevents notary flow recursion
+	if !bytes.Equal(nr.MainTransaction.Scripts[1].InvocationScript, p.dummyInvocationScript) {
+		return nil, ErrTXAlreadyHandled
+	}
+
+	currentAlphabet, err := p.alphaKeys()
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch Alphabet public keys: %w", err)
+	}
+
+	err = p.validateCosigners(nr.MainTransaction.Signers, currentAlphabet)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate main TX's notary attribute
+	err = p.validateAttributes(nr.MainTransaction.Attributes, currentAlphabet)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate main TX's witnesses
+	err = p.validateWitnesses(nr.MainTransaction.Scripts, currentAlphabet)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate main TX expiration
+	err = p.validateExpiration(nr.FallbackTransaction)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		opCode opcode.Opcode
+		param  []byte
+	)
+
+	ctx := vm.NewContext(nr.MainTransaction.Script)
+	ops := make([]Op, 0, 16) // 16 is maximum num of opcodes for calling contracts with 4 args(no arrays of arrays)
+
+	for {
+		opCode, param, err = ctx.Next()
+		if err != nil {
+			return nil, fmt.Errorf("could not get next opcode in script: %w", err)
+		}
+
+		if opCode == opcode.RET {
+			break
+		}
+
+		ops = append(ops, Op{code: opCode, param: param})
+	}
+
+	opsLen := len(ops)
+
+	// check if it is tx with contract call
+	if !bytes.Equal(ops[opsLen-1].param, p.contractSysCall) {
+		return nil, errNotContractCall
+	}
+
+	// retrieve contract's script hash
+	contractHash, err := util.Uint160DecodeBytesBE(ops[opsLen-2].param)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode contract hash: %w", err)
+	}
+
+	// retrieve contract's method
+	contractMethod := string(ops[opsLen-3].param)
+
+	err = p.validateParameterOpcodes(ops[:opsLen-3])
+	if err != nil {
+		return nil, err
+	}
+
+	return parsedNotaryEvent{
+		hash:       contractHash,
+		notaryType: NotaryTypeFromString(contractMethod),
+		params:     ops[:opsLen-3],
+		raw:        nr,
+	}, nil
+}
+
+func (p Preparator) validateParameterOpcodes(oo []Op) error {
+	// check for unexpected NeoVM opcodes
+	for _, o := range oo {
+		switch {
+		// only PUSH(and PACK for arrays) codes are allowed;
+		// number of params and their content must be checked
+		// in a notary parser and a notary handler of a
+		// particular contract
+		case o.code <= opcode.PUSH16 || o.code == opcode.PACK:
+		default:
+			return errUnexpectedOpcode
+		}
+	}
+
+	return nil
+}
+
+func (p Preparator) validateExpiration(fbTX *transaction.Transaction) error {
+	if len(fbTX.Attributes) != 3 || fbTX.Attributes[1].Type != transaction.NotValidBeforeT {
```

Comment:
The attributes order is undefined in general, so here you check the way neo-go RPC client constructs them. So I suggest to keep attributes length check and use `fbTX.GetAttributes(transation.NotValidBefore)` instead of fetching attribute by index.

Reply:
thanks, added more general attribute check